cranelift/fuzzgen/src/lib.rs

```diff
@@ -137,11 +137,18 @@ where
         })
     }
 
-    fn generate_test_inputs(&mut self, signature: &Signature) -> Result<Vec<TestCaseInput>> {
-        let num_tests = self.u.int_in_range(self.config.test_case_inputs.clone())?;
-        let mut inputs = Vec::with_capacity(num_tests);
+    fn generate_test_inputs(mut self, signature: &Signature) -> Result<Vec<TestCaseInput>> {
+        let mut last_len = usize::MAX;
+        let mut inputs = Vec::new();
+
+        // Continue generating input as long as we just consumed some of self.u. Otherwise we'll
+        // generate the same test input again on the next iteration, and there's no point testing
+        // it twice. Note that once self.u becomes empty we naturally don't consume any more of it,
+        // so this check covers that case as well as the case of testing a no-argument function.
+        // This lets us consume all remaining fuzz data.
+        while self.u.len() < last_len {
+            last_len = self.u.len();
```

Comment:
That's how I wrote it first, actually! I could go back to `loop` plus a conditional `break` at the end if you want. But just testing `u.is_empty` doesn't work if generating an input doesn't consume any of `u`: it loops forever then. The usual way that would happen is if the generated function doesn't take any parameters, but I haven't ruled out the possibility that a parameter could only have one legal value and therefore consume no input. So instead I check that the loop makes some progress on every iteration, and stop after the first iteration where it doesn't make progress.

I agree that using `usize::MAX` here is subtle. We'd get nearly the same result from initializing `last_len` to `self.u.len() + 1`, if that would make you feel better?

Reply:
I guess we could also have a maximum length on `inputs` as well.

Otherwise I guess I would prefer making `last_len` an option and doing something like

```rust
while last_len.map_or(true, |last_len| self.u.len() < last_len) {
    ...
}
```

Reply:
I'm definitely not a fan of that alternative :laughing: but I can still move the test to the end of the loop, if that works for you.